### PR TITLE
[CFP-235] Migrate external users tables to Design System

### PR DESCRIPTION
**Pull request by naseberry**
#### What
Migrate external users table markup to the GOVUK Design System Table markup.

#### Ticket
[Migrate external users table markup to the GOVUK Design System Table](https://dsdmoj.atlassian.net/browse/CFP-235)

#### Why
Part of a larger change to:

- continue migration to GOVUK Frontend
- resolve accessibility issues identified in the 2020 DAC report
- reduce then remove deprecated code

#### Routes
 - [x] /external_users
 - [x] /external_users/claims
 - [x] /external_users/admin/external_users/:id
 - [x] /external_users/admin/providers/:id
 - [x] /external_users/claims/:id
 - [x] /external_users/claims/:id/summary
 - [x] /external_users/claims/archived
 - [x] /external_users/claims/authorised
 - [x] /external_users/claims/outstanding
 - [x] /help/cookie-details
 - [x] api/release_notes

#### TO-DO
- [x] clean up dead CSS styling
- [x] sanity check on `staging`

### CODE REVIEW

File: app/views/external_users/claims/_main_claims_list.html.haml [hunk 1]
```diff
@@ -1,97 +1,99 @@
...
+        = sortable 'case_number', t('.case_number'), 'aria-label': t('.case_number_sort_label'), id: 'case_number_ms'
+        - if current_user.persona.admin?
+          = sortable 'advocate', t('.advocate_litigator'), 'aria-label': t('.advocate_sort_label'), id: 'advocate_ms'
+        = sortable 'total_inc_vat', t('.total'), 'aria-label': t('.total_inc_vat_sort_label'), id: 'total_inc_vat_ms'
+        = sortable 'amount_assessed', t('.assessed'), 'aria-label': t('.amount_assessed_sort_label'), id: 'amount_assessed_ms'
+        = sortable 'state', t('.status'), 'aria-label': t('.state_sort_label'), id: 'state_ms'
+        = sortable 'last_submitted_at', t('.submission_date'), 'aria-label': t('.last_submitted_at_sort_label'), id: 'last_submitted_at_ms'
+
+    - present_collection(claims).each do |claim|
+      = govuk_table_row(id: dom_id(claim), class: claim.state) do
```
Comment (imtiazAhmd) [inline app/views/external_users/claims/_main_claims_list.html.haml:58]: Note to self: Needs encapsulation with a single quote!

File: app/views/external_users/claims/_prosecution_evidence_summary.html.haml [hunk 2]
```diff
@@ -1,7 +1,6 @@
-%table{ class: 'summary' }
-  %tbody
-    %tr
-      %th{ scope: 'row', class: 'bold' }
-        = t("common.prosecution_evidence")
-      %td
-        = t(claim.prosecution_evidence?.class)
+%dl.govuk-summary-list
```
Comment (naseberry) [inline app/views/external_users/claims/_prosecution_evidence_summary.html.haml:1]: You may wish to use the SummaryListHelpers for this component

File: app/views/external_users/claims/_prosecution_evidence_summary.html.haml [hunk 3]
```diff
@@ -1,7 +1,6 @@
...
+  .govuk-summary-list__row
+    %dt.govuk-summary-list__key
+      = t("common.prosecution_evidence")
+    %dd.govuk-summary-list__value
+      = t(claim.prosecution_evidence?.class)
```
Comment (naseberry) [inline app/views/external_users/claims/_prosecution_evidence_summary.html.haml]: You may wish to use the SummaryListHelpers for this component

File: app/views/case_workers/admin/allocations/_re_allocation.html.haml [hunk 4]
```diff
@@ -29,7 +29,7 @@
           = f.govuk_submit t('.re_allocate'), name: 'tab', value: 'allocated'
 
       - if @claims.any?
-        = govuk_table(class: 'report js-checkbox-table') do
+        = govuk_table(class: 'js-checkbox-table') do
```
Comment (naseberry) [inline app/views/case_workers/admin/allocations/_re_allocation.html.haml:32]: Is the class `.js-checkbox-table` still relevant?

File: app/views/external_users/claims/supporting_documents/_existing.html.haml [hunk 5]
```diff
@@ -1,28 +1,26 @@
 - if @claim.documents.any?(&:persisted?)
-  %table.previously-uploaded
-    %caption
+  = govuk_table(class: 'previously-uploaded') do
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_existing.html.haml]: Is the class `.previously-uploaded` still relevant?

File: app/views/external_users/claims/supporting_documents/_new.html.haml [hunk 6]
```diff
@@ -16,17 +16,17 @@
...
         = t('.file_name')
-      %th{ scope: 'col' }
+      = govuk_table_th do
         = t('.status')
-      %th{ scope: 'col' }
+      = govuk_table_th do
         = t('.action')
 
-  %tbody
+  %tbody.govuk-table__body
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_new.html.haml]: `%tbody` needs to be replaced by `= govuk_table_tbody do`

File: app/views/geckoboard_api/widgets/_stats_table.html.haml [hunk 7]
```diff
@@ -2,12 +2,15 @@
...
-      %th= source
+      = govuk_table_th do
+        = "Source"
   - payload.to_a.each do |row|
-    %tr
+    = govuk_table_row do
       - row.each do |value|
-        %td= value
+        = govuk_table_td('data-label': value) do
+          = value
```
Comment (naseberry) [inline app/views/geckoboard_api/widgets/_stats_table.html.haml]: Semantics: table missing both `thead` and `tbody`

File: app/views/case_workers/admin/allocations/_re_allocation.html.haml (hunk 4 above)
Comment (imtiazAhmd) [reply app/views/case_workers/admin/allocations/_re_allocation.html.haml:32]: its a selector in `app/webpack/javascripts/modules/Modules.TableRowClick.js`

Comment (imtiazAhmd) [reply app/views/case_workers/admin/allocations/_re_allocation.html.haml:32]: Should I change it to an id ?

File: app/views/external_users/claims/supporting_documents/_new.html.haml (hunk 6 above)
Comment (imtiazAhmd) [reply app/views/external_users/claims/supporting_documents/_new.html.haml]: this is the end of line code in that haml. rest of the table is rendered by`app/webpack/javascripts/modules/external_users/claims/Dropzone.js`.

If changed to table helper it throws error, as there is nothing after `do`

File: app/webpack/javascripts/modules/external_users/claims/Dropzone.js [hunk 8]
```diff
@@ -140,7 +140,7 @@ moj.Modules.Dropzone = {
     formData.append('document[document]', file)
 
     const tableBody = $('.files tbody')
```
[the remark this answers is not in the document]
Comment (imtiazAhmd) [reply app/webpack/javascripts/modules/external_users/claims/Dropzone.js:142]: its a selector, I think it is looking for `tbody` element within a div with class `files`. If you think it is required to mention the class of this `tbody` as-well, I will add it.

File: app/views/external_users/claims/_prosecution_evidence_summary.html.haml (hunk 3 above)
Comment (imtiazAhmd) [reply app/views/external_users/claims/_prosecution_evidence_summary.html.haml]: Updated

File: app/views/external_users/claims/supporting_documents/_existing.html.haml (hunk 5 above)
Comment (imtiazAhmd) [reply app/views/external_users/claims/supporting_documents/_existing.html.haml]: Updated

File: app/views/geckoboard_api/widgets/_stats_table.html.haml (hunk 7 above)
Comment (imtiazAhmd) [reply app/views/geckoboard_api/widgets/_stats_table.html.haml]: Updated

File: app/views/case_workers/admin/allocations/_re_allocation.html.haml (hunk 4 above)
Comment (imtiazAhmd) [reply app/views/case_workers/admin/allocations/_re_allocation.html.haml:32]: removed from scss

File: app/views/external_users/claims/supporting_documents/_new.html.haml (hunk 6 above)
Comment (imtiazAhmd) [reply app/views/external_users/claims/supporting_documents/_new.html.haml]: Updated

File: app/views/case_workers/admin/allocations/_allocation.html.haml [hunk 9]
```diff
@@ -42,7 +42,7 @@
   %h2.govuk-heading-m
     = t('.allocation_queue')
 
-  = govuk_table(class: 'report app-jq-datatable', style: 'width:100%') do
+  = govuk_table(class: 'app-jq-datatable', style: 'width:100%') do
```
Comment (jsugarman) [inline app/views/case_workers/admin/allocations/_allocation.html.haml]: nit 
```suggestion
  = govuk_table(class: 'app-jq-datatable', style: 'width: 100%') do
```

File: app/views/external_users/claims/_final_claim_cost_summary.html.haml [hunk 10]
```diff
@@ -1,136 +1,136 @@
...
-      %th.numeric{ scope: 'col' }
+      = govuk_table_th_numeric do
         = t('shared.summary.total_amount')
 
-  %tbody
+  = govuk_table_tbody do
     - if claim.fixed_fee_case?
-      %tr.fx-seed-fixedFees{ data: { autovat: @claim.apply_vat? ? 'true' : 'false', seed: claim.raw_fixed_fees_total } }
-        %td
+      = govuk_table_row(class: 'fx-seed-fixedFees', data: { autovat: @claim.apply_vat? ? 'true' : 'false', seed: claim.raw_fixed_fees_total }) do
```
Comment (jsugarman) [inline app/views/external_users/claims/_final_claim_cost_summary.html.haml:18]: 🤔  are these `fx-seed-` classes needed on this page? They are use by `SideBar.js`  to reflect changes in the fees to my knowledge but this page is just a summary and so amounts cannot be changed anyway.

Anyway, maybe safer to leave in

File: app/views/external_users/claims/supporting_documents/_summary.html.haml [hunk 11]
```diff
@@ -2,50 +2,50 @@
...
-    %caption.govuk-visually-hidden
+  = govuk_table do
+    = govuk_table_caption(class: 'govuk-visually-hidden') do
       = "Evidence supplied on disk"
-    %tbody
-      %tr
-        %th{ scope:'row', class: 'bold' }
+    = govuk_table_tbody do
+      = govuk_table_row do
+        = govuk_table_th(scope:'row', class: 'bold') do
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_summary.html.haml]: Class `.bold` is no longer required, as the `th` element now has the relevant font weight

File: app/views/external_users/claims/supporting_documents/_summary.html.haml [hunk 12]
```diff
@@ -2,50 +2,50 @@
...
+= govuk_table do
+  = govuk_table_caption(class: 'govuk-visually-hidden') do
     = t('.caption')
-  %tbody
+  = govuk_table_tbody do
     - claim.documents.each do |document|
-      %tr
-        %th.summary-headings{scope:'row'}
+      = govuk_table_row do
+        = govuk_table_th(scope:'row', class: 'summary-headings') do
```
Comment (naseberry) [inline app/views/external_users/claims/supporting_documents/_summary.html.haml]: Is class `.summary-headings` still required?

File: app/views/external_users/claims/_main_claims_list.html.haml [hunk 13]
```diff
@@ -1,97 +1,99 @@
...
-          %td.messages{ 'data-label': t('.messages') }
-            - if claim.messages.any?
-              = govuk_link_to claim.unread_messages_for(current_user).any? ? t('.view_with_messages', message_count: claim.unread_messages_for(current_user).count) : t('.view'), "#{external_users_claim_path(claim, messages: true)}#messages", 'aria-label': t('.view_messages_label', case_number: claim.case_number)
-            - else
-              %span.none= t('.none')
+
+  = govuk_table_tbody do
+    = govuk_table_row(class: 'mobile-sort') do
+      = govuk_table_th do
+        Sort by:
```
Comment (jsugarman) [inline app/views/external_users/claims/_main_claims_list.html.haml:48]: **nit**
while you are here you could move this to a translation

File: app/webpack/stylesheets/_tables.scss [hunk 14]
```diff
@@ -1,425 +0,0 @@
-// stylelint-disable max-nesting-depth, selector-class-pattern, selector-no-qualifying-type
```
Comment (jsugarman) [inline app/webpack/stylesheets/_tables.scss:1]: woohoo!

File: features/page_objects/claim_summary_page.rb [hunk 15]
```diff
@@ -1,7 +1,7 @@
 class ClaimSummaryPage < BasePage
   set_url "/external_users/claims/{claim_id}/summary"
 
-  element :continue, ".form-buttons > .govuk-button:nth-of-type(1)"
+  element :continue, "main .govuk-button-group > .govuk-button:nth-of-type(1)"
```
Comment (jsugarman) [inline features/page_objects/claim_summary_page.rb:4]: **nit** do we need to scope it to `main`?

```suggestion
  element :continue, ".govuk-button-group > .govuk-button:nth-of-type(1)"
```

File: features/step_definitions/common_steps.rb [hunk 16]
```diff
@@ -125,7 +125,7 @@
 Then(/^Claim '(.*?)' should be listed with a status of '(.*?)'(?: and a claimed amount of '(.*?)')?$/) do |case_number, status, claimed|
   my_claim = @external_user_home_page.claim_for(case_number)
   expect(my_claim).not_to be_nil
-  expect(my_claim.state.text).to eq(status.upcase)
+  expect(my_claim.state.text).to include(status.upcase)
```
Comment (jsugarman) [inline features/step_definitions/common_steps.rb]: has it changed to return an array?

**Review by jsugarman — COMMENTED** [review]
👍  looks good. 

but will sanity test now

**Review by jsugarman — APPROVED** [review]
Have sanity tested on staging and can't find any problems so am approving but might be worth getting others to do more extensive sanity testing.

File: features/page_objects/claim_summary_page.rb (hunk 15 above)
Comment (imtiazAhmd) [reply features/page_objects/claim_summary_page.rb:4]: Yes. when this page starts rendering, it shows cookie banner that has the same button group.

File: features/step_definitions/common_steps.rb (hunk 16 above)
Comment (imtiazAhmd) [reply features/step_definitions/common_steps.rb]: @naseberry

File: app/views/external_users/claims/_main_claims_list.html.haml (hunk 13 above)
Comment (imtiazAhmd) [reply app/views/external_users/claims/_main_claims_list.html.haml:48]: We discussed about moving all of these to a translation files. And decided to work on it in a separate piece of work as this PR is getting too long.

File: features/step_definitions/common_steps.rb (hunk 16 above)
Comment (naseberry) [reply features/step_definitions/common_steps.rb]: No it hasn't, this should not have been committed, therefore I've now removed it.

**Comment by jrmhaig** [comment]
I tried the drop zone for the evidence documents, as I saw that there are changes in the javascript. This works OK. I noticed a difference in the document list, though;

## Before

Screenshot 2021-11-10 at 11 42 45

## After

Screenshot 2021-11-10 at 11 44 11

(The new table is the full width of the filename, which goes off the page by some way).

Are we able to get this wrapped? Or truncated so that it fits on the page?

**Review by jrmhaig — APPROVED** [review]

**Comment by imtiazAhmd** [comment]
> I tried the drop zone for the evidence documents, as I saw that there are changes in the javascript. This works OK. I noticed a difference in the document list, though;
> 
> ## Before
> Screenshot 2021-11-10 at 11 42 45
> 
> ## After
> Screenshot 2021-11-10 at 11 44 11
> 
> (The new table is the full width of the filename, which goes off the page by some way).
> 
> Are we able to get this wrapped? Or truncated so that it fits on the page?

Resolved

**Pull request merged**